src/main/java/frc/robot/subsystems/Led.java

```diff
@@ -17,26 +19,88 @@
     private int frame = 0;
     private int timer = 0;
 
+    public enum BallColor {
+        RED(Color.kFirstRed), BLUE(Color.kFirstBlue), NONE(new Color(0, 0, 0));
+
+        private final Color color;
+
+        private BallColor(Color color) {
+            this.color = color;
+        }
+
+        public Color get() {
+            return color;
+        }
+    }
+
     public Led(final LedMap map) {
         led = map.getLed();
         ledBuffer = map.getLedBuffer();
         led.setLength(ledBuffer.getLength());
         led.start();
     }
 
+    private void runAnimation(LightAnimation animation, double brightness) {
+        if (timer % 5 == 0) {
+            for (int i = 0; i < ledBuffer.getLength(); i++) {
+                Color c = animation.getColor(frame,
+                        i / (ledBuffer.getLength() / 10));
+                ledBuffer.setLED(i, new Color(
+                        c.red * brightness,
+                        c.green * brightness,
+                        c.blue * brightness));
+            }
+            led.setData(ledBuffer);
+            frame++;
+        }
+    }
+
     public CommandBase animate(LightAnimation animation, double brightness) {
         return running("Animate", () -> {
-            if (timer % 5 == 0) {
+            runAnimation(animation, brightness);
+            timer++;
+        });
+    }
+
+    private BallColor matchColor(Color c) {
+        if (c.red > c.green && c.red > c.blue) {
+            return BallColor.RED;
+        }
+        if (c.blue > c.green && c.blue > c.red) {
+            return BallColor.BLUE;
+        }
+        return BallColor.NONE;
+    }
+
+    public CommandBase showBallColors(ColorBufferSupplier colorBuffer, LightAnimation defaultAnimation,
+            double brightness) {
+        return running("Show Ball Colors", () -> {
+            SampleBuffer<Color> colors = colorBuffer.getAsColorBuffer();
+            if (colors.size() == 0) {
+                runAnimation(defaultAnimation, brightness);
+            } else {
                 for (int i = 0; i < ledBuffer.getLength(); i++) {
-                    Color c = animation.getColor(frame,
-                            i / (ledBuffer.getLength() / 10));
+                    double sin = (Math.sin(timer / 10.0) + 1.0) / 2.0;
+                    Color c;
+
+                    if (i < ledBuffer.getLength() / 2) {
```

Comment:
Flip this logic - instead of doing the calculations for each LED, do the calculations once per ball, then divide the LED strip into a number of segments equal to the number of balls. Then it'd be really easy to apply the same color to each LED in the segment.